File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.cpp
class Trie {
public:
    Trie* children[26];
    bool isTerminal;
    Trie() {
        for(auto i=0;i<26;i++) children[i]=NULL;
        isTerminal = false;
    }
    
    void insert(string word) {
        auto cur = this;
        for(auto i:word){
            if(!cur->children[i-'a']){
                cur->children[i-'a'] = new Trie();
            }
            cur = cur->children[i-'a'];
        }
        
        cur->isTerminal = true;
    }
    
    bool search(string word) {
        auto cur = this;
        for(auto i:word){
            if(!cur->children[i-'a']){
                return false;
            }
            cur = cur->children[i-'a'];
        }
        
        return cur->isTerminal;
    }
    
    bool startsWith(string prefix) {
        auto cur = this;
        for(auto i:prefix){
            if(!cur->children[i-'a']){
                return false;
            }
            cur = cur->children[i-'a'];
        }
        
        bool childrenExist = false;
        for(auto i:cur->children){
            if(i) {
                childrenExist = true;
                break;
            }
        }
        
        return childrenExist||cur->isTerminal;
        
    }
};

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.py
TR = lambda : defaultdict(TR)
WORD = "word"
class Trie:

    def __init__(self):
        self.root = TR()

    def insert(self, word: str) -> None:
        tmp = self.root
        for c in word:
            if c not in tmp:
                tmp[c] = TR()
            tmp = tmp[c]
        tmp[WORD]=True

    def search(self, word: str) -> bool:
        tmp = self.root
        for c in word:
            if c not in tmp: return False
            tmp = tmp[c]
        return WORD in tmp

    def startsWith(self, prefix: str) -> bool:
        tmp = self.root
        for c in prefix:
            if c not in tmp: return False
            tmp = tmp[c]
        return len(tmp.keys())>0


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)